src/store.c

```diff
@@ -339,47 +407,104 @@ static void writer_close(
+        inverted->data[x].key = pairs->data[x].val;
+    }
+
+    inverted->len = pairs->len;
+    inverted->cap = pairs->cap;
+
+    return inverted;
+}
+
 bool rill_store_write(
         const char *file,
-        rill_ts_t ts, size_t quant,
+        rill_ts_t ts,
+        size_t quant,
         struct rill_pairs *pairs)
 {
     rill_pairs_compact(pairs);
     if (!pairs->len) return true;
 
+    struct rill_pairs* inverted = flip_values(pairs);
     struct vals *vals = vals_from_pairs(pairs);
+    struct vals *invert_vals = vals_from_pairs(inverted);
+
     if (!vals) goto fail_vals;
 
     struct rill_store store = {0};
-    if (!writer_open(&store, file, vals, pairs->len, ts, quant)) {
+    if (!writer_open(&store, file, vals, invert_vals,
+                     pairs->len, inverted->len, ts, quant)) {
         rill_fail("unable to create '%s'", file);
         goto fail_open;
     }
 
     struct indexer *indexer = indexer_alloc(pairs->len);
-    struct encoder coder = writer_begin(&store, vals, indexer);
+    struct indexer *indexer_b = indexer_alloc(inverted->len);
+
+    store.head->index_b_off = store.head->index_off + indexer_cap(pairs->len);
+    store.head->data_off    = store.head->index_b_off + indexer_cap(inverted->len);
+
+    store.index   = (void *) ((uintptr_t) store.vma + store.head->index_off);
+    store.index_b = (void *) ((uintptr_t) store.vma + store.head->index_b_off);
+    store.data    = (void *) ((uintptr_t) store.vma + store.head->data_off);
+
+    struct encoder coder =
+        store_encoder(&store,
+                      indexer,
+                      vals,
+                      store.head->data_off);
 
     for (size_t i = 0; i < pairs->len; ++i) {
         if (!coder_encode(&coder, &pairs->data[i])) goto fail_encode;
     }
     if (!coder_finish(&coder)) goto fail_encode;
 
-    store.head->keys = coder.keys;
+    store.head->data_b_off = store.head->data_off + coder_off(&coder);
+    store.data_b = (void *) ((uintptr_t) store.vma + store.head->data_b_off);
+
+    struct encoder coder_b =
+        store_encoder(&store,
+                      indexer_b,
+                      invert_vals,
+                      store.head->data_b_off);
+
+    for (size_t i = 0; i < inverted->len; ++i) {
+        if (!coder_encode(&coder_b, &inverted->data[i])) goto fail_encode;
+    }
+    if (!coder_finish(&coder_b)) goto fail_encode;
+
+    writer_flush_indices(&store, indexer, indexer_b);
+
+    store.head->keys  = coder.keys  + coder_b.keys;
     store.head->pairs = coder.pairs;
 
-    writer_close(&store, indexer, store.head->data_off + coder_off(&coder));
+    writer_close(&store, store.head->data_b_off + coder_off(&coder_b));
 
     coder_close(&coder);
+    coder_close(&coder_b);
+
     indexer_free(indexer);
+    indexer_free(indexer_b);
+
     free(vals);
+    free(inverted);
     return true;
 
   fail_encode:
     coder_close(&coder);
-    writer_close(&store, indexer, 0);
+    coder_close(&coder_b);
```

Comment:
You can end up closing `coder_b` before it's even created which will do bad things as you're dealing with uninitialised memory. To solution is to be very liberal with the introduction of labels.

Reply:
I think I fixed most of these, though I'll be going over them again, as I don't feel comfortable with my graceful cleanup skills yet.